File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionString.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.util.Map;

public class TransformerFunctionString extends TransformerFunction {
    public TransformerFunctionString() {
        super(FunctionDescription.of(
            Map.of(
            "json", ArgumentType.of(ArgType.Boolean).position(0).defaultBoolean(false)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var value = context.getUnwrapped(null);
        if (context.getBoolean("json")) {
            // although gson.toJson will return "null" eventually, this is quicker
            if (value == null) {
                return "null";
            }
            var adapter = context.getAdapter();
            return adapter.toString(value);
        }
        return context.getAsString(value);
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionString.ts
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { getAsString } from "../JsonHelpers";
import { ArgType } from "./common/ArgType";
import { JSONBig } from "./common/FunctionHelpers";

class TransformerFunctionString extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        json: { type: ArgType.Boolean, position: 0, defaultBoolean: false },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const value = await context.getUnwrapped(null);
    if (await context.getBoolean("json")) {
      return JSONBig.stringify(value);
    }
    return getAsString(value);
  }
}

export default TransformerFunctionString;
